Sources/NIOSOCKS/Messages/ClientRequest.swift

```diff
@@ -0,0 +1,182 @@
+
+// MARK: - AddressType
+
+/// The address used to connect to the target host.
+public struct AddressType: Hashable {
+    
+    public var address: SocketAddress
+    
+    /// How many bytes are needed to represent the address
+    public var size: Int {
+        switch address {
+        case .v4:
+            return 4
+        case .v6:
+            return 16
+        case .unixDomainSocket:
+            fatalError("Unsupported")
+        }
+    }
+    
+    public init(address: SocketAddress) {
+        self.address = address
+    }
+}
+
+extension ByteBuffer {
+    
+    mutating func readAddresType() throws -> AddressType? {
+        let save = self
+        guard let type = self.readInteger(as: UInt8.self) else {
+            self = save
+            return nil
+        }
+        
+        switch type {
+        case 0x01:
+            return try self.readIPv4Address()
+        case 0x03:
+            return try self.readDomain()
+        case 0x04:
+            return try self.readIPv6Address()
+        default:
+            throw InvalidAddressType(actual: type)
+        }
+    }
+    
+    mutating func readIPv4Address() throws -> AddressType? {
+        let save = self
+        guard
+            let bytes = self.readBytes(length: 4),
+            let port = try self.readPort()
+        else {
+            self = save
+            return nil
+        }
+        return .init(address: try .init(packedIPAddress: ByteBuffer(bytes: bytes), port: port))
+    }
+    
+    mutating func readIPv6Address() throws -> AddressType? {
+        let save = self
+        guard
+            let bytes = self.readBytes(length: 16),
+            let port = try self.readPort()
+        else {
+            self = save
+            return nil
+        }
+        return .init(address: try .init(packedIPAddress: ByteBuffer(bytes: bytes), port: port))
+    }
+    
+    mutating func readDomain() throws -> AddressType? {
+        let save = self
+        guard
+            let length = self.readInteger(as: UInt8.self),
+            let bytes = self.readBytes(length: Int(length)),
+            let port = try self.readPort()
+        else {
+            self = save
+            return nil
+        }
+        let host = String(decoding: bytes, as: Unicode.UTF8.self)
+        return .init(address: try .makeAddressResolvingHost(host, port: port))
+    }
+    
+    mutating func readPort() throws -> Int? {
+        return self.readInteger(as: UInt16.self).flatMap { Int($0)}
+    }
+    
+    @discardableResult mutating func writeAddressType(_ type: AddressType) -> Int {
+        switch type.address {
+        case .v4(let address):
+            return self.writeInteger(UInt8(1))
+                + self.writeInteger(address.address.sin_addr.s_addr, endianness: .little)
+                + self.writeInteger(address.address.sin_port, endianness: .little)
+        case .v6(let address):
+            let (p1, p2, p3, p4) = address.address.sin6_addr.__u6_addr.__u6_addr32
```

Comment:
The odds of this being cross-platform are basically zero. Unfortunately we temporarily need to drop down to unsafe code here because we don't have a good way to provide a packed representation of the address, but for now I think an extension on ByteBuffer to write a `sockaddr_in6` is the way to go here. To write that, you can use `withUnsafeBytes(of: address.address.sin6_addr)`.